File: telegram-bot-game/opus-4-5/bot.js
require('dotenv').config();
const TelegramBot = require('node-telegram-bot-api');

const token = process.env.TELEGRAM_BOT_TOKEN;

if (!token) {
  console.error('Error: TELEGRAM_BOT_TOKEN environment variable is not set');
  process.exit(1);
}

const bot = new TelegramBot(token, { polling: true });

// Store game state per chat
const games = new Map();

// Default range
const DEFAULT_MIN = 1;
const DEFAULT_MAX = 100;

function getGame(chatId) {
  return games.get(chatId);
}

function createGame(chatId, min = DEFAULT_MIN, max = DEFAULT_MAX) {
  const secretNumber = Math.floor(Math.random() * (max - min + 1)) + min;
  const game = {
    secretNumber,
    min,
    max,
    attempts: 0,
    active: true
  };
  games.set(chatId, game);
  return game;
}

function formatRange(game) {
  return `${game.min}–${game.max}`;
}

// /start command
bot.onText(/\/start/, (msg) => {
  const chatId = msg.chat.id;
  const game = createGame(chatId);

  bot.sendMessage(chatId,
    `Welcome to the Number Guessing Game!\n\n` +
    `I'm thinking of a number between ${formatRange(game)}.\n` +
    `Try to guess it! Just type a number.\n\n` +
    `Commands:\n` +
    `/newgame - Start a new game\n` +
    `/range <min> <max> - Set custom range (e.g., /range 1 1000)\n` +
    `/help - Show help`
  );
});

// /help command
bot.onText(/\/help/, (msg) => {
  const chatId = msg.chat.id;

  bot.sendMessage(chatId,
    `Number Guessing Game - Help\n\n` +
    `How to play:\n` +
    `1. I pick a random number in a range\n` +
    `2. You guess by sending a number\n` +
    `3. I'll tell you if the answer is higher or lower\n` +
    `4. Keep guessing until you find it!\n\n` +
    `Commands:\n` +
    `/start - Start the bot and a new game\n` +
    `/newgame - Start a new game with current range\n` +
    `/range <min> <max> - Set custom range and start new game\n` +
    `  Example: /range 1 1000\n` +
    `/help - Show this help message`
  );
});

// /newgame command
bot.onText(/\/newgame/, (msg) => {
  const chatId = msg.chat.id;
  const existingGame = getGame(chatId);
  const min = existingGame?.min || DEFAULT_MIN;
  const max = existingGame?.max || DEFAULT_MAX;

  const game = createGame(chatId, min, max);

  bot.sendMessage(chatId,
    `New game started!\n` +
    `I'm thinking of a number between ${formatRange(game)}.\n` +
    `Make your guess!`
  );
});

// /range command
bot.onText(/\/range(?:\s+(\d+)\s+(\d+))?/, (msg, match) => {
  const chatId = msg.chat.id;

  if (!match[1] || !match[2]) {
    const game = getGame(chatId);
    const currentRange = game ? formatRange(game) : `${DEFAULT_MIN}–${DEFAULT_MAX}`;
    bot.sendMessage(chatId,
      `Current range: ${currentRange}\n\n` +
      `Usage: /range <min> <max>\n` +
      `Example: /range 1 1000`
    );
    return;
  }

  const min = parseInt(match[1], 10);
  const max = parseInt(match[2], 10);

  if (min >= max) {
    bot.sendMessage(chatId,
      `Invalid range! The minimum must be less than the maximum.\n` +
      `Example: /range 1 1000`
    );
    return;
  }

  if (max - min < 1) {
    bot.sendMessage(chatId,
      `Range too small! Please use a larger range.`
    );
    return;
  }

  const game = createGame(chatId, min, max);

  bot.sendMessage(chatId,
    `Range updated to ${formatRange(game)}!\n` +
    `New game started. Make your guess!`
  );
});

// Handle guesses (any message that's a number)
bot.on('message', (msg) => {
  const chatId = msg.chat.id;
  const text = msg.text;

  // Ignore commands
  if (!text || text.startsWith('/')) {
    return;
  }

  // Check if input is a valid number
  const guess = parseInt(text.trim(), 10);

  if (isNaN(guess)) {
    bot.sendMessage(chatId,
      `Please enter a valid number, or use /help for commands.`
    );
    return;
  }

  const game = getGame(chatId);

  if (!game || !game.active) {
    bot.sendMessage(chatId,
      `No active game! Use /newgame to start playing.`
    );
    return;
  }

  // Check if guess is in range
  if (guess < game.min || guess > game.max) {
    bot.sendMessage(chatId,
      `Please guess a number between ${formatRange(game)}.`
    );
    return;
  }

  game.attempts++;

  if (guess < game.secretNumber) {
    bot.sendMessage(chatId, `Higher! Try again.`);
  } else if (guess > game.secretNumber) {
    bot.sendMessage(chatId, `Lower! Try again.`);
  } else {
    // Winner!
    game.active = false;
    const attempts = game.attempts;
    const attemptWord = attempts === 1 ? 'attempt' : 'attempts';

    bot.sendMessage(chatId,
      `Congratulations! You guessed it!\n\n` +
      `The number was ${game.secretNumber}.\n` +
      `You found it in ${attempts} ${attemptWord}!\n\n` +
      `Want to play again? Use /newgame or /range to start a new game.`
    );
  }
});

// Handle polling errors
bot.on('polling_error', (error) => {
  console.error('Polling error:', error.message);
});

console.log('Number Guessing Bot is running...');

File: telegram-bot-game/gemini-3-pro/bot.py
import logging
import random
import os
import sys
from typing import Dict, Any

from telegram import Update, ReplyKeyboardMarkup, ReplyKeyboardRemove
from telegram.ext import (
    ApplicationBuilder,
    ContextTypes,
    CommandHandler,
    MessageHandler,
    filters,
)

# Configure logging
logging.basicConfig(
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
    level=logging.INFO
)

# Game state storage: chat_id -> game_data
# game_data structure:
# {
#   'target': int,
#   'attempts': int,
#   'min': int,
#   'max': int,
#   'active': bool
# }
games: Dict[int, Dict[str, Any]] = {}

DEFAULT_MIN = 1
DEFAULT_MAX = 100

async def start(update: Update, context: ContextTypes.DEFAULT_TYPE):
    """Sends a welcome message and starts a game."""
    user_name = update.effective_user.first_name
    await update.message.reply_text(
        f"Hello {user_name}! Welcome to the Number Guessing Game.\n"
        f"I'm thinking of a number between {DEFAULT_MIN} and {DEFAULT_MAX}.\n"
        "Type /newgame to start or /help for more info."
    )
    # Automatically start a game with default settings if not exists
    await start_new_game(update, context)

async def help_command(update: Update, context: ContextTypes.DEFAULT_TYPE):
    """Displays help information."""
    help_text = (
        "🎮 *Number Guessing Game Help* 🎮\n\n"
        "/start - Welcome message and start default game\n"
        "/newgame - Start a new game with current range\n"
        "/range <min> <max> - Set a new custom range (e.g., /range 1 500)\n"
        "/help - Show this message\n\n"
        "Just type a number to make a guess!"
    )
    await update.message.reply_text(help_text, parse_mode='Markdown')

async def set_range(update: Update, context: ContextTypes.DEFAULT_TYPE):
    """Sets a custom range for the game."""
    chat_id = update.effective_chat.id
    args = context.args

    if len(args) != 2:
        await update.message.reply_text("Usage: /range <min> <max> (e.g., /range 1 1000)")
        return

    try:
        min_val = int(args[0])
        max_val = int(args[1])

        if min_val >= max_val:
            await update.message.reply_text("Invalid range: Minimum must be less than Maximum.")
            return

        # Initialize or update state preferences
        if chat_id not in games:
            games[chat_id] = {}
        
        games[chat_id]['min'] = min_val
        games[chat_id]['max'] = max_val
        
        # Start a new game with these settings immediately
        await start_new_game(update, context)

    except ValueError:
        await update.message.reply_text("Please provide valid integers for the range.")

async def start_new_game(update: Update, context: ContextTypes.DEFAULT_TYPE):
    """Initializes a new game state."""
    chat_id = update.effective_chat.id
    
    # Get user preferences or defaults
    min_val = games.get(chat_id, {}).get('min', DEFAULT_MIN)
    max_val = games.get(chat_id, {}).get('max', DEFAULT_MAX)

    target = random.randint(min_val, max_val)
    
    games[chat_id] = {
        'target': target,
        'attempts': 0,
        'min': min_val,
        'max': max_val,
        'active': True
    }

    await update.message.reply_text(
        f"🎲 New game started! Guess a number between {min_val} and {max_val}."
    )

async def handle_guess(update: Update, context: ContextTypes.DEFAULT_TYPE):
    """Handles text input as a guess."""
    chat_id = update.effective_chat.id
    game = games.get(chat_id)

    if not game or not game.get('active'):
        await update.message.reply_text("No active game. Type /newgame to start one!")
        return

    user_text = update.message.text.strip()

    # Handle non-integer input
    if not user_text.lstrip('-').isdigit():
        await update.message.reply_text("That's not a valid number. Please enter an integer.")
        return

    guess = int(user_text)
    game['attempts'] += 1
    
    if guess < game['target']:
        await update.message.reply_text("🔼 Higher!")
    elif guess > game['target']:
        await update.message.reply_text("🔽 Lower!")
    else:
        # Correct guess
        attempts = game['attempts']
        game['active'] = False
        await update.message.reply_text(
            f"🎉 Correct! The number was {game['target']}.\n"
            f"You found it in {attempts} attempt{'s' if attempts != 1 else ''}.\n\n"
            "Type /newgame to play again!"
        )

if __name__ == '__main__':
    token = os.environ.get("TELEGRAM_BOT_TOKEN")
    if not token:
        print("Error: TELEGRAM_BOT_TOKEN environment variable not set.")
        sys.exit(1)

    application = ApplicationBuilder().token(token).build()

    application.add_handler(CommandHandler("start", start))
    application.add_handler(CommandHandler("newgame", start_new_game))
    application.add_handler(CommandHandler("help", help_command))
    application.add_handler(CommandHandler("range", set_range))
    
    # Filter for text messages that are not commands
    application.add_handler(MessageHandler(filters.TEXT & (~filters.COMMAND), handle_guess))

    print("Bot is running...")
    application.run_polling()
